### ISSUE TITLE
DefaultJedisClientConfig password exposure issue

### ISSUE THREAD

**Issue by ivanfrias** [redis/jedis#4021]
The class DefaultJedisClientConfig overrides the getPassword method that returns a String.
Returning a string might be considered a potential security issue since an attacker might inspect the heap and find the value in plaintext.
Ideally we should just pass-through the value supplied by the provider [here](https://github.com/redis/jedis/blob/master/src/main/java/redis/clients/jedis/DefaultJedisClientConfig.java#L79) and not create a String based on the char[] array.

### Expected behavior

Return a char[] instead of String

### Actual behavior

A string is returned.

### Steps to reproduce:
N/A

### Redis / Jedis Configuration
N/A

#### Jedis version:
N/A

#### Java version:
N/A

**Comment by sazzad16** [comment]
`String getPassword()` is still there to support legacy applications without breaking. We may remove it at some point in favor of `getCredentialsProvider()`.

**Comment by sazzad16** [comment]
Tagged 6.0.0 just so it stays in front of eyes more.

**Comment by wickdynex** [comment]
If possible, I would like to take ownership of this issue. In my opinion, I can replace `String` with `char[]` and modify the  `DefaultJedisClientConfig` class and `DefaultRedisCredentials` methods. However, I have three questions:

1. Using `char[]` instead of `String` is for easier password clearing. When exactly should we clear the password? Or is it not necessary to clear it?
2. I couldn't find any unit tests for `getPassword`. Do I need to write a unit test for it to follow TDD development practices?
3. Which branch should I work on?

Looking forward to ur reply.

**Comment by sazzad16** [comment]
@wickdynex Thank you for your interest.

I would ask you not to work on this specific issue right now. I have already mentioned in my earlier https://github.com/redis/jedis/issues/4021#issuecomment-2468128171,

> We may remove it (`String getPassword()`) at some point in favor of `getCredentialsProvider()`.

The concerned users should use RedisCredentials/RedisCredentialsProvider.

**Comment by sazzad16** [comment]
> So is this issue have assigned to your teammate?

No. No one.

**Comment by wickdynex** [comment]
> > So is this issue have assigned to your teammate?
> 
> No. No one.

If u'ld like to assign to me, I’d be happy to take on this issue. 
Here’s the approach I plan to take to fix it:
- Use `DefaultJedisClientConfig.getCredentialsProvider()`method to replace `DefaultJedisClientConfig .getPassword()` where it used at. And refactor the usage.

Would you agree with this approach? Please let me know if you have any feedback or suggestions.

#### Followed Questions:
1. Should I write unit tests for this method? I noticed that the project may be lacking tests in some areas, and it could be a good idea to add tests if necessary.
2. Which branch should I work on?

Looking forward to ur reply.

**Comment by sazzad16** [comment]
> Should I write unit tests for this method?

Ideally, yes.

> Which branch should I work on?

`master`.

**Comment by sazzad16** [comment]
> Please assign this issue to me.

@wickdynex It's not necessary. You're welcome to craft a PR about any improvement even without the assignment. Thanks.

**Comment by wickdynex** [comment]
> > Please assign this issue to me.
> 
> @wickdynex It's not necessary. You're welcome to craft a PR about any improvement even without the assignment. Thanks.

Thank u. I'll try `craft PR`.☺️

**Comment by wickdynex** [comment]
> > Should I write unit tests for this method?
> 
> Ideally, yes.
> 
> > Which branch should I work on?
> 
> `master`.

I’ve found that not only the `DefaultJedisClientConfig` class, but also the `JedisCluster`, `JedisFactory`, `JedisPool`, and `JedisPooled` classes are using `String` instead of `char[]` to store the password, along with related tests. So terrible! This means that refactoring this issue will require a rather large PR. Do I need to refactor all instances of this password-related bug, or should I just focus on refactoring `DefaultJedisClientConfig` and modify the related code to use `password.toCharArray()` instead?
Please give me some advice, thanks. Looking forward to ur reply.

**Comment by aarifkhan7** [comment]
Hey @wickdynex, are you still working on this issue?  If you're busy or not actively working on it, I’d be happy to help.

**Comment by wickdynex** [comment]
> Hey [@wickdynex](https://github.com/wickdynex), are you still working on this issue? If you're busy or not actively working on it, I’d be happy to help.

Thanks for ur help, I'm glad to find u leave a comment here. Now I'm busy with other Issues, because this Issue is so hard to fix, the code are coupling together, too headache. I used to try, but finally gived up. I need to express my appreciation again.😀

**Comment by wickdynex** [comment]
@ggivo 
> What do you think about limiting the scope to `DefaultJedisClientConfig` .

In my opinion, It might be a good way to fix it. This issue is a typical "technical debt", if u don't fix it early, it'll become a huge problem. I used to fix it, failed😵Only way I think is also to refactor this Class. I approve.

**Comment by aarifkhan7** [comment]
@ggivo I agree with your approach to this problem. From what I understand, we will introduce a new method in JedisClientConfig (the interface for DefaultJedisClientConfig) with a signature like char[] getPasswordAsChars(), while marking the existing getPassword() method as @Deprecated.

In future iterations of this fix, we can gradually update the constructors to use the new method from JedisClientConfig and deprecate the older constructors. This will help fixing this issue while ensuring backward compatibility. 

Am I right?

@ggivo @wickdynex Your guidance on this approach would be greatly appreciated.

**Comment by ggivo** [comment]
@aarifkhan7 no need for new method as per https://github.com/redis/jedis/issues/4021#issuecomment-2468128171 it shoudl be deprecated in favor of using `getCredentialsProvider()`
